### PR TITLE
add subscribe/unsubscribe for C++ client

### PR DESCRIPTION
**Pull request by vicentebolea**
It changes the way the message are read in the following way:

 1. A polling thread will be filling share queues of incoming messages
 2. When send is called, it will block until its reply message is
    deposited in the shared queue.

A more detailed description is in this file:

https://github.com/Kitware/wslink/blob/77902aa45201b1820a5969e654295fa17f81b99b/tests/chat-rpc-pub-sub/clients/cpp/README.md

### CODE REVIEW

File: cpp/wsWebsocketConnection.h [hunk 1]
```diff
@@ -348,28 +367,33 @@ class wsWebsocketConnection
...
+      while (QueueRPCMessages.empty()) {
+        CVRPCMessages.wait(lock);
+      }
 
-    // The make_printable() function helps print a ConstBufferSequence
-    if (this->Debug)
-    {
-      std::cout << beast::buffers_to_string(buffer.data()) << std::endl;
+      *result = QueueRPCMessages.front();
+      QueueRPCMessages.pop();
```
Comment (jourdain) [inline cpp/wsWebsocketConnection.h:378]: Why do you keep the front and remove the last?

**Comment by jourdain** [comment]
It looks like you can only have 1 subscription on a given topic. That is probably fine, but we may want to add a comment to mention that limitation.

File: tests/chat-rpc-pub-sub/clients/cpp/cli_test.cc [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  ws.subscribe(TOPIC, 
+      [](const json& json) -> void 
+      {
+        cout << json["id"] << " -> " << json["result"] << endl;
+      });
+
+  std::string line;
+  while (std::cin >> line)
+  {
+    json args = { line.c_str() };
```
Comment (jourdain) [inline tests/chat-rpc-pub-sub/clients/cpp/cli_test.cc:29]: We may want to exit the loop with a command like `:q` or `exit`. Just an idea.

File: cpp/wsWebsocketConnection.h (hunk 1 above)
Comment (vicentebolea) [reply cpp/wsWebsocketConnection.h:378]: Unlike the `std::list` or `std::stack`, the `std::queue::pop` deletes the first element.

File: tests/chat-rpc-pub-sub/clients/cpp/cli_test.cc (hunk 2 above)
Comment (vicentebolea) [reply tests/chat-rpc-pub-sub/clients/cpp/cli_test.cc:29]: Yes that would make it more usable, currently it just exits with either `Ctrl-D` or `Ctrl-C`.

Comment (jourdain) [reply tests/chat-rpc-pub-sub/clients/cpp/cli_test.cc:29]: Let's do it as it is a small change and easy to understand.

File: cpp/wsWebsocketConnection.h [hunk 3]
```diff
@@ -395,12 +421,81 @@ class wsWebsocketConnection
...
+          }
+        }
+        else if (type == "rpc")
+        {
+          std::unique_lock<std::mutex> lock (MutexRPCMessages);
+          QueueRPCMessages.push(result);
+          CVRPCMessages.notify_all();
+
+        } else {
+          std::cerr << "unsupported type of message: " << type << std::endl;
```
[the remark this answers is not in the document]
Comment (jourdain) [reply cpp/wsWebsocketConnection.h:474]: system might be for initial handshake and maybe for attachments.

**Pull request merged**